### PR TITLE
Design doc for #1818: Issue #1818 Design: Make sure synthetic names are unambiguously identifiable

### PR DESCRIPTION
**Pull request by johnynek**
Design doc.

Refs #1818

### CODE REVIEW

File: docs/design/1818-make-sure-synthetic-names-are-unambiguously-identifiable.md [hunk 1]
```diff
@@ -0,0 +1,225 @@
...
+
+This keeps protobuf schema unchanged and preserves compatibility for valid existing artifacts. Legacy artifacts that already lost information due prior collisions are treated as best-effort decode.
+
+## Problem statement
+
+Today synthetic names are represented as `Identifier.Name("_...")` and distinguished by convention.
+
+That creates three concrete problems:
+
+1. Identifier equality and ordering are keyed by `asString` only, so `Name("_x")` and `Backticked("_x")` can collide in `Map`/`Set`/`SortedMap` paths even though they represent different source-level identities.
```
Comment (johnynek) [inline docs/design/1818-make-sure-synthetic-names-are-unambiguously-identifiable.md]: I think current code may leverage this, especially with operators. I think we leverage in the compiler that `operator +` and `Backticked("+")` are `==`.

I think to be fully compatible we would need do something like:
```scala
def eqv(a: Identifier, b: Identifier) =
  (a, b) match {
    case (Synthetic(sa), Synthetic(sb)) => sa == sb
    case (Synthetic(_), _) | (_, Synthetic(_)) => false
    case (notSynA, notSynB) => notSynA.asString == notSynB.asString
  }
```
Or something like that.

Then if protobuf serialization uses `sourceCodeRepr` then we should be okay, because `_foo` won't parse as a operator, backticked, or name.

A similar technique can be used to implement `Order[Bindable]` and `Hash[Bindable]` which allows us to consider all the syntactic forms equal as they were before.

Re-read the relevant code and see adjust this doc as needed.

Comment (johnynek) [reply docs/design/1818-make-sure-synthetic-names-are-unambiguously-identifiable.md]: Updated the design doc to match this compatibility constraint. In the "Make identifier identity synthetic-aware and backward compatible" section, the proposal now uses a hybrid rule: `Synthetic` is disjoint from non-synthetics, while non-synthetic identifiers keep existing `asString`-based equality/order/hash behavior (including current operator/backticked interop). I also updated the implementation plan, tests, acceptance criteria, risks, and rollout notes to reflect that exact boundary.

**Comment by johnynek** [comment]
Addressed the review feedback by editing `docs/design/1818-make-sure-synthetic-names-are-unambiguously-identifiable.md` to preserve non-synthetic compatibility while isolating synthetics. Ran required pre-push check `scripts/test_basic.sh`; it passed.
